--- autoconf/config.guess	2019-11-12 18:24:50.946913000 -0500
+++ autoconf/config.guess	2019-11-12 18:25:38.510143957 -0500
@@ -969,8 +969,8 @@
 	  *)    echo hppa-unknown-linux-gnu ;;
 	esac
 	exit ;;
-    ppc64:Linux:*:*)
-	echo powerpc64-unknown-linux-gnu
+    ppc64le:Linux:*:*)
+	echo powerpc64le-unknown-linux-gnu
 	exit ;;
     ppc:Linux:*:*)
 	echo powerpc-unknown-linux-gnu
--- isl/config.guess	2019-11-12 18:24:51.051072000 -0500
+++ isl/config.guess	2019-11-12 18:25:50.713724628 -0500
@@ -969,8 +969,8 @@
 	  *)    echo hppa-unknown-linux-gnu ;;
 	esac
 	exit ;;
-    ppc64:Linux:*:*)
-	echo powerpc64-unknown-linux-gnu
+    ppc64le:Linux:*:*)
+	echo powerpc64le-unknown-linux-gnu
 	exit ;;
     ppc:Linux:*:*)
 	echo powerpc-unknown-linux-gnu
--- osl/autoconf/config.guess	2019-11-12 18:24:51.124881000 -0500
+++ osl/autoconf/config.guess	2019-11-12 18:26:31.090769461 -0500
@@ -969,8 +969,8 @@
 	  *)    echo hppa-unknown-linux-gnu ;;
 	esac
 	exit ;;
-    ppc64:Linux:*:*)
-	echo powerpc64-unknown-linux-gnu
+    ppc64le:Linux:*:*)
+	echo powerpc64le-unknown-linux-gnu
 	exit ;;
     ppc:Linux:*:*)
 	echo powerpc-unknown-linux-gnu
