--- config.guess	2016-07-01 16:25:50.000000000 -0400
+++ config.guess	2020-07-08 13:55:05.813568501 -0400
@@ -918,8 +918,8 @@ EOF
     ppc:Linux:*:*)
 	echo powerpc-unknown-linux-gnu
 	exit ;;
-    ppc64:Linux:*:*)
-	echo powerpc64-unknown-linux-gnu
+    ppc64le:Linux:*:*)
+	echo powerpc64le-unknown-linux-gnu
 	exit ;;
     alpha:Linux:*:*)
 	case `sed -n '/^cpu model/s/^.*: \(.*\)/\1/p' < /proc/cpuinfo` in
--- easel/devkit/sqc	2016-07-01 16:25:53.000000000 -0400
+++ easel/devkit/sqc	2020-07-08 13:54:19.990841203 -0400
@@ -507,6 +507,8 @@ foreach $tmpfile (keys(%used_tmpfile)) {
 unlink $tmp if -e $tmp;
 unlink "$tmp.stderr" if -e "$tmp.stderr";
 
+# Set exit code based on how many tests failed
+exit ($badtest > 0);
 
 sub check_status
 {
--- hmmer/src/impl_sse/p7_oprofile.c	2016-07-01 16:26:27.000000000 -0400
+++ hmmer/src/impl_sse/p7_oprofile.c	2020-07-08 13:52:48.227492967 -0400
@@ -670,7 +670,7 @@ biased_byteify(P7_OPROFILE *om, float sc
   uint8_t b;
 
   sc  = -1.0f * roundf(om->scale_b * sc);                          /* ugh. sc is now an integer cost represented in a float...           */
-  b   = (sc > 255 - om->bias_b) ? 255 : (uint8_t) sc + om->bias_b; /* and now we cast, saturate, and bias it to an unsigned char cost... */
+  b   = (sc > 255 - om->bias_b) ? 255 : (uint8_t)(signed int) sc + om->bias_b; /* and now we cast, saturate, and bias it to an unsigned char cost... */
   return b;
 }
  
@@ -686,7 +686,7 @@ unbiased_byteify(P7_OPROFILE *om, float
   uint8_t b;
 
   sc  = -1.0f * roundf(om->scale_b * sc);       /* ugh. sc is now an integer cost represented in a float...    */
-  b   = (sc > 255.) ? 255 : (uint8_t) sc;	/* and now we cast and saturate it to an unsigned char cost... */
+  b   = (sc > 255.) ? 255 : (uint8_t)(signed int)sc;	/* and now we cast and saturate it to an unsigned char cost... */
   return b;
 }
  
